src/lightning/fabric/strategies/xla_fsdp.py

```diff
@@ -0,0 +1,403 @@
+            if len(modules) == 0:
+                raise ValueError(
+                    "Could not find a XLAFSDP model in the provided checkpoint state. Please provide the model as"
+                    " part of the state like so: `load_checkpoint(..., state={'model': model, ...})`. Make sure"
+                    " you set up the model (and optimizers if any) through the strategy before loading the checkpoint."
+                )
+            if len(modules) > 1:
+                raise ValueError(
+                    "Found multiple XLAFSDP modules in the given state. Loading checkpoints with FSDP is"
+                    " currently limited to a single model per checkpoint. To load multiple models, call the"
+                    " load method for each model separately with a different path."
+                )
+
+            _, module = list(modules.items())[0]
+            sharded_ckpt = torch.load(file)
+
+            module.load_state_dict(sharded_ckpt["model"], strict=strict)
+            for opt_key, opt in optimizers.items():
+                opt.load_state_dict(sharded_ckpt[opt_key])
+
+            # Load anything leftover from sharded_ckpt
+            loaded_metadata_keys = sharded_ckpt.keys() - modules.keys() - optimizers.keys()
+            requested_metadata_keys = state.keys() - modules.keys() - optimizers.keys()
+            _validate_keys_for_strict_loading(requested_metadata_keys, loaded_metadata_keys, strict=strict)
+            for key in requested_metadata_keys:
+                if key in loaded_metadata_keys:
+                    state[key] = sharded_ckpt[key]
+                    loaded_metadata_keys.remove(key)
+
+            metadata = {}
+            if len(loaded_metadata_keys):
+                for key in loaded_metadata_keys:
+                    metadata[key] = sharded_ckpt[key]
+
+            # remove "shard_metadata" that is loaded in
+            if "shard_metadata" in metadata:
+                metadata.pop("shard_metadata")
+
+            return metadata
+        elif self._state_dict_type == "full":
+            file = os.path.join(path, "checkpoint_consolidated.pth")
+            if not Path(file).is_file():
+                raise ValueError(
+                    f"The path {str(file)!r} does not point to a valid full checkpoint. Make sure the path points to a"
+                    " directory with a full XLAFSDP checkpoint."
+                )
+            rank_zero_warn(
+                "Loading a full checkpoint will only load the full model."
+                " Optimizer and any additional metadata are not included."
+            )
+            if len(modules) > 0:
+                raise ValueError(
+                    "Found a XLAFSDP model in the provided checkpoint state."
+                    " Please provide the model without any XLAFSDP wrapper."
+                )
+            full_ckpt = torch.load(str(file))
+            state["model"].load_state_dict(full_ckpt["model"], strict=strict)
```

Comment:
We should add logic like https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/strategies/fsdp.py#L601-L611 so that we return the proper leftover state in the full state dict branch

Reply:
So, the full checkpoint that is saved only saves the model, ignoring the optimizer and other metadata. Note: checkpoint consolidation uses https://github.com/pytorch/xla/blob/master/torch_xla/distributed/fsdp/state_dict_utils.py#L109.

Because of this limitation, should we leave as is?

Reply:
Then I would suggest:

```suggestion
            state["model"].load_state_dict(full_ckpt.pop("model"), strict=strict)
            return full_ckpt
```

Which will show the leftover state